#code to creat a taxonomic tree
#last edtied Dec 12, 2023 by A. R. Martinig

```{r setup, include = FALSE}
knitr::opts_chunk$set(echo = TRUE)


options(scipen=999, dplyr.width = Inf, tibble.print_min = 50, repos='http://cran.rstudio.com/') #scipen forces outputs to not be in scientific notation #dplyr.width will show all columns for head() function and tibble.print_min sets how many rows are printed and repos sets the cran mirror

#load libraries

#to install ggtree and ggtreeExtra for the first time
#if (!requireNamespace("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")
#BiocManager::install("ggtree")

#if (!requireNamespace("BiocManager", quietly=TRUE))
#    install.packages("BiocManager")
#BiocManager::install("ggtreeExtra")

packages=c("googlesheets4", "tidyverse", "rotl", "dplyr", "ape", "ggplot2", "ggnewscale", "ggtree", "ggtreeExtra") 

# Install packages not yet installed
installed_packages <- packages %in% rownames(installed.packages())
if (any(installed_packages == FALSE)) {
  install.packages(packages[!installed_packages])
}

# Packages loading
lapply(packages, library, character.only = TRUE)

select<-dplyr::select



```

## R Markdown

```{r load data}


# Specify the URL of your Google Sheet
url <- "https://docs.google.com/spreadsheets/d/1ZWQUCG5wuLIPJbhYfRW_rvFVb4qFiWtFm2e0BZ5WBkE/edit#gid=1046848762"

# Read the Google Sheet into a dataframe
Species_info <- read_sheet(url) %>%
	filter(!composite_variable=="Y", !obsID=="TBD") %>%
    #slice(210:250) %>%
  select(c("reference", "paperID", "common_name", "species")) %>%
    mutate(species = case_when(species == "Thryomanes bewickii, Junco hyemalis, Pipilo maculatus, Melospiza melodia" ~ "Passeriformes", #went with one of the species that is not in the tree yet
  		TRUE ~ species))

```


```{r make tree}

#str(Species_info)
length(unique(Species_info$species)) #148
unique(Species_info$species) #needs cleaning
table(Species_info$species)

#Species_info$species <- gsub("_"," ", Species_info$species) #get rid of the underscores

# Make a table of how many words each name has:
table((lengths(gregexpr("\\W+", Species_info$species)) + 1)) #34 records with >2 words species names (subspecies etc.)
 
 # Count the number of words in the 'species' column
word_counts <- lengths(gregexpr("\\S+", Species_info$species))

# Identify records with more than two words
more_than_two_words <- Species_info[word_counts > 2, ]

# Display the result
print(more_than_two_words)
 
taxa <- tnrs_match_names(unique(Species_info$species)) #fails due to non-recognizable names

# Fix obviously incompatible names (subspecies, name variants):
Species_info <- Species_info %>%
  mutate(
  animal = 
  	case_when(
  		#species == "Spermophilus parryii plesius" ~ "Spermophilus parryii",
  		#species %in% c("Strix occidentalis lucida", "Strix occidentalis caurina")  ~ "Strix occidentalis",
  		#species == "Canis lupus arabs" ~ "Canis lupus",
  		#species == "Lagopus lagopus scoticus" ~ "Lagopus lagopus",
  		#species == "Branta bernicla nigricans" ~ "Branta bernicla",
  		species == "Vulpes macrotis mutica" ~ "Vulpes macrotis",
  		#species == "Panthera tigris tigris" ~ "Panthera tigris",
  		#species == "Gorilla beringei beringei" ~ "Gorilla beringei",
  		species == "Macaca mulatta tcheliensis" ~ "Macaca mulatta",
  		#species == "Gorilla gorilla gorilla" ~ "Gorilla gorilla",
  		species == "Calidris alpina schinzii" ~ "Calidris alpina",
  		#species == "Homo sapiens sapiens" ~ "Homo sapiens",
  		species == "Passeriformes" ~ "Junco hyemalis", #went with one of the species that is not in the tree yet
  		TRUE ~ species))

#NA -> drop rows with "NA" as species name
Species_info <- Species_info[Species_info$animal != "NA", ]

table(Species_info$animal, useNA = "always") #no NA
length(unique(Species_info$animal)) #148

taxa <- tnrs_match_names(unique(Species_info$animal)) #runs fine
names(taxa)
synonyms(taxa) ##list of all synonyms
taxa$unique_name # main TOL names
table(taxa$approximate_match) ##2 approximate matches
taxa %>% filter(approximate_match==TRUE)
table(taxa$flags) # flags 26 names with problems (hidden, hybrid, incertae_sedis_inherited, infraspecific, infraspecific) - will need fixing

mytree <- tol_induced_subtree(ott_ids = taxa$ott_id, label_format= "name") 
#this will fail with: Error: HTTP failure: 400


## re-run taxa matching and tree retrieval:
taxa <- tnrs_match_names(unique(Species_info$animal)) #runs fine
table(taxa$approximate_match) #2 approximate matches
table(taxa$flags) #flags 26 things - infraspecific and sibling_higher
mytree <- tol_induced_subtree(ott_ids = taxa$ott_id, label_format= "name") #this now works fine

plot(mytree, show.tip.label = T, cex = 0.8, no.margin = TRUE)

is.binary(mytree) #if it returns FALSE that means you have several branches coming from the same node

mytree$tip.label

## Tree tip labels need some cleaning:
mytree$tip.label <- gsub("\\(.*", "", mytree$tip.label) #remove comments
mytree$tip.label <- gsub("_"," ", mytree$tip.label) #get rid of the underscores
mytree$tip.label <- trimws(mytree$tip.label) #getting rid of the trailing whitespace

mytree$tip.label

#this lets me check which species names don't match up
#you want to get 100% matching records between the two (your original labels and the labels in the tree)
sort(intersect(as.character(mytree$tip.label), unique(Species_info$species))) ## 130 names are matching -  got quite a few left to fix
sort(setdiff(unique(Species_info$species), as.character(mytree$tip.label))) ## names from species column not matching with tip label
sort(setdiff(as.character(mytree$tip.label), unique(Species_info$species))) ## names from tip.label not matching in animal column

#need to fix any of the substitutions AND labels that are matched, but incorrectly

## Reverse substitutions of species names in the tree tip labels:
#mytree$tip.label[mytree$tip.label == "Strix occidentalis"] <- "Strix occidentalis lucida & S. o. caurina"
#mytree$tip.label[mytree$tip.label == "Branta bernicla"] <- "Branta bernicla nigricans" 
#mytree$tip.label[mytree$tip.label == "Vulpes macrotis"] <- "Vulpes macrotis mutica" 
#mytree$tip.label[mytree$tip.label == "Gorilla beringei"] <- "Gorilla beringei beringei" 
mytree$tip.label[mytree$tip.label == "Macaca mulatta"] <- "Macaca mulatta tcheliensis" 
#mytree$tip.label[mytree$tip.label == "Gorilla gorilla"] <- "Gorilla gorilla gorilla" 
mytree$tip.label[mytree$tip.label == "Calidris alpina"] <- "Calidris alpina schinzii" 
#mytree$tip.label[mytree$tip.label == "Homo sapiens"] <- "Homo sapiens sapiens" 
mytree$tip.label[mytree$tip.label == "Junco hyemalis"] <- "Passeriformes"
mytree$tip.label[mytree$tip.label == "Canis lupus arctos"] <- "Canis lupus arabs"
		
## Other substitutions of species names made in the tree tip labels:
mytree$tip.label[mytree$tip.label == "Urocitellus parryii plesius"] <- "Spermophilus parryii plesius"

#check matching records again
sort(intersect(as.character(mytree$tip.label), unique(Species_info$species))) ## 135 names are matching - not yet all fixed 










plot(mytree, show.tip.label = T, cex = 0.8, no.margin = TRUE)
str(mytree) #335 tips

write.tree(mytree, file = "species_tree.tre") #save the tree 
# mytree <- read.tree(file = "species_tree.tre") #if you need to read the tree

write.csv(Species_info, file = "Species_info.csv")

```

```{r plot tree}

# count of papers per species name:
sample_data <- count(Species_info,animal)
 
sample_data$animal %in% mytree$tip.label
mytree$tip.label %in% sample_data$animal
 
ggtree(mytree, layout="circular") +                    
   geom_tiplab(                             # adds name of sample to tip of its branch 
     color = 'black',                       
     offset = 1,
     size = 1,
     geom = "text",
     align = TRUE) 
 
ggtree(mytree, layout = "circular", branch.length = 'none') %<+% sample_data + # %<+% adds dataframe with sample data to tree 
   new_scale_color() +                             # allows to add an additional color scheme for another variable
     geom_tippoint(
       mapping = aes(color = n),  
       size = 1) +
   scale_colour_gradient(low = "yellow", high = "red") +
   geom_tiplab(                             # adds name of sample to tip of its branch 
     color = 'black',                       
     offset = 1,
     size = 1,
     geom = "text",
     align = TRUE) 
   
 
ggtree(mytree, layout="circular", branch.length = "none") # simple circular tree with all tips aligned
ggtree(mytree, layout="fan", open.angle=10, yscale = "none")
 
ggsave("species_tree_circular_v0.pdf", width = 12, height = 10, scale = 10)
 
p
```

```{r plot tree}

sample_data <- count(Species_info, animal)
sample_data <- as.data.frame(sample_data)
sample_data2 <-mutate(sample_data, tip.label = animal)
mytree$tip.label <- strip_ott_ids(mytree$tip.label, remove_underscores = TRUE)
 summary(sample_data2$tip.label)
 sample_data2$count <- as.numeric(sample_data2$count)
sample_data3 <- left_join(sample_data2, Species_info[,c("species_name","broad_taxa")], by = "species_name")
sample_data4 <- distinct(sample_data3, animal, .keep_all = TRUE)
 is.na(sample_data4)
 
sample_data4$animal <- as.factor(sample_data4$animal)
sample_data4$tip.label <- as.factor(sample_data4$tip.label)
sample_data4$broad_taxa<- as.factor(sample_data4$broad_taxa)

#logging n
sample_data4$n_logged <- log(sample_data4$n)


cols <- c("yellow", "violet", "turquoise", "tomato", "thistle", "springgreen", "navy", "orange", "lightgoldenrod", "lightpink", "khaki", "lightslateblue", "red", "seagreen", "plum", "hotpink4", "moccasin", "olivedrab")

tree2 <- ggtree(mytree, layout = "circular", lwd = 0.75) %<+% sample_data4 + aes(col = broad_taxa)+ theme(legend.position = "bottom") + scale_colour_manual(values = cols, (title = "Broad taxa")) 

tree2


tree3 <- tree2 + new_scale_fill() + geom_fruit(geom = geom_bar, mapping = aes(x = log(n), col = "gray30"), stat = "identity", col = "gray30", orientation = "y", axis.params = list(axis = "x",text.angle = -45, hjust = 0, text.size = 3), grid.params = list(alpha = 0.35),offset = 0.085, pwidth = 0.55, alpha = 0.8) + guides(fill ="none") 

tree3

count(sample_data4, broad_taxa)
count(sample_data4, species_name)

## will add silhouettes to most common species
##need to manually change log axis back to original numbers

```