#code to creat a taxonomic tree figure
#last edtied Jan 8, 2024 by A. R. Martinig



packages=c("rotl", "ape", "ggplot2", "ggnewscale", "ggtree", "ggtreeExtra") 

# Install packages not yet installed
installed_packages <- packages %in% rownames(installed.packages())
if (any(installed_packages == FALSE)) {
  install.packages(packages[!installed_packages])
}

# Packages loading
lapply(packages, library, character.only = TRUE)

select<-dplyr::select


#read in data and edit as needed
Species_info <- df %>%
  select(c("reference", "paperID", "common_name", "species", "species_class", "dispersal_type", "fitness_higher_level")) %>%
    mutate(species = 
    	case_when(
    		species == "Thryomanes bewickii, Junco hyemalis, Pipilo maculatus, Melospiza melodia" ~ "Passeriformes", #went with one of the species that is not in the tree yet
    		species %in% c("Canis lupus", "Canis lupus arabs") ~ "Canis lupus & Canis lupus arabs", 
    		species %in% c("Lagopus lagopus", "Lagopus lagopus scoticus")  ~ "Lagopus lagopus & Lagopus lagopus scoticus", 
    			species %in% c("Panthera tigris", "Panthera tigris tigris")  ~ "Panthera tigris & Panthera tigris tigris", 
  		TRUE ~ species)) %>%
  	group_by(species)

head(Species_info)
#str(Species_info)
length(unique(Species_info$species)) #139
unique(Species_info$species) #needs cleaning
table(Species_info$species)

#Species_info$species <- gsub("_"," ", Species_info$animal) #get rid of the underscores

# Make a table of how many words each name has:
table((lengths(gregexpr("\\W+", Species_info$species)) + 1)) #48 records with >2 words species names (subspecies etc.)
 
 # Count the number of words in the 'species' column
word_counts <- lengths(gregexpr("\\S+", Species_info$species))

# Identify records with more than two words
more_than_two_words <- Species_info[word_counts > 2, ]

# Display the result
print(more_than_two_words)
 
taxa <- tnrs_match_names(unique(Species_info$species)) #fails due to non-recognizable names

# Fix obviously incompatible names (subspecies, name variants):
Species_info <- Species_info %>%
  mutate(
  animal = 
  	case_when(
  		species == "Vulpes macrotis mutica" ~ "Vulpes macrotis",
  		species == "Macaca mulatta tcheliensis" ~ "Macaca mulatta",
  		species == "Calidris alpina schinzii" ~ "Calidris alpina",
  		species == "Passeriformes" ~ "Junco hyemalis", #went with one of the species that is not in the tree yet
  		species == "Canis lupus & Canis lupus arabs" ~ "Canis lupus", 
    	species == "Lagopus lagopus & Lagopus lagopus scoticus"  ~ "Lagopus lagopus", 
    	species == "Panthera tigris & Panthera tigris tigris"  ~ "Panthera tigris", 
  		TRUE ~ species))
  		

#NA -> drop rows with "NA" as species name
Species_info <- Species_info[Species_info$animal != "NA", ]

table(Species_info$animal, useNA = "always") #no NA
length(unique(Species_info$animal)) #139

taxa <- tnrs_match_names(unique(Species_info$animal)) #runs fine
names(taxa)
synonyms(taxa) ##list of all synonyms
taxa$unique_name # main TOL names
table(taxa$approximate_match) ##1 approximate match
taxa %>% filter(approximate_match==TRUE)
table(taxa$flags) # flags 21 names with problems (hidden, hybrid, incertae_sedis_inherited, infraspecific, infraspecific) - will need fixing

mytree <- tol_induced_subtree(ott_ids = taxa$ott_id, label_format= "name") 
#this will fail


plot(mytree, show.tip.label = T, cex = 0.8, no.margin = TRUE)

is.binary(mytree) #if it returns FALSE that means you have several branches coming from the same node

mytree$tip.label

## Tree tip labels need some cleaning:
mytree$tip.label <- gsub("\\(.*", "", mytree$tip.label) #remove comments
mytree$tip.label <- gsub("_"," ", mytree$tip.label) #get rid of the underscores
mytree$tip.label <- trimws(mytree$tip.label) #getting rid of the trailing whitespace

mytree$tip.label

#this lets me check which species names don't match up
#you want to get 100% matching records between the two (your original labels and the labels in the tree)
sort(intersect(as.character(mytree$tip.label), unique(Species_info$species))) ## 124 names are matching -  got quite a few left to fix
sort(setdiff(unique(Species_info$species), as.character(mytree$tip.label))) ## names from species column not matching with tip label
sort(setdiff(as.character(mytree$tip.label), unique(Species_info$species))) ## names from tip.label not matching in animal column

#need to fix any of the substitutions AND labels that are matched, but incorrectly

## Reverse substitutions of species names in the tree tip labels:
mytree$tip.label[mytree$tip.label == "Calidris alpina"] <- "Calidris alpina schinzii"
mytree$tip.label[mytree$tip.label == "Canis lupus"] <- "Canis lupus & Canis lupus arabs" 
mytree$tip.label[mytree$tip.label == "Junco hyemalis"] <- "Passeriformes" 
mytree$tip.label[mytree$tip.label == "Vulpes macrotis"] <- "Vulpes macrotis mutica" 
mytree$tip.label[mytree$tip.label == "Macaca mulatta"] <- "Macaca mulatta tcheliensis" 
mytree$tip.label[mytree$tip.label == "Panthera tigris"] <- "Panthera tigris & Panthera tigris tigris" 
mytree$tip.label[mytree$tip.label == "Lagopus lagopus scoticus"] <- "Lagopus lagopus & Lagopus lagopus scoticus" #red grouse

		
## Other substitutions of species names made in the tree tip labels:
mytree$tip.label[mytree$tip.label == "Urocitellus parryii plesius"] <- "Spermophilus parryii plesius" #Arctic ground squirrel
mytree$tip.label[mytree$tip.label == "Urocitellus columbianus"] <- "Spermophilus columbianus" #Columbian ground squirrel


mytree$tip.label[mytree$tip.label == "Urocitellus townsendii"] <- "Spermophilus townsendii" #Townsend's Ground Squirrel
mytree$tip.label[mytree$tip.label == "Mustela putorius furo"] <- "Mustela furo" #ferret
mytree$tip.label[mytree$tip.label == "Vulpes lagopus"] <- "Alopex lagopus" #arctic fox
mytree$tip.label[mytree$tip.label == "Myodes glareolus"] <- "Clethrionomys glareolus" #bank vole
mytree$tip.label[mytree$tip.label == "Myodes rufocanus"] <- "Clethrionomys rufocanus" #Grey red-backed vole
mytree$tip.label[mytree$tip.label == "Poecile palustris"] <- "Parus palustris" #marsh tit
mytree$tip.label[mytree$tip.label == "Setophaga petechia"] <- "Dendroica petechia" #yellow warbler



#check matching records again
sort(intersect(as.character(mytree$tip.label), unique(Species_info$species))) ## 139 names are matching - all fixed 

plot(mytree, show.tip.label = T, cex = 0.8, no.margin = TRUE)
str(mytree) #276 tips

head(Species_info)

#these tip labels are only to be used to relabel for the final plot, not for the data that is being shared

d.tree <- mutate(Species_info, tip.label = species)  # For the correspondence between tree and data
d.tree$tip.label = as.factor(d.tree$tip.label)  # Convert to factor
d.tree$tip.label <- gsub("_", " ", d.tree$tip.label)  # remove underscores from data
mytree$tip.label <- strip_ott_ids(mytree$tip.label, remove_underscores = TRUE)  # remove underscores from tree

summary(d.tree$tip.label)  # Need to remove duplicate species for building the tree

 
 
d.tree <- d.tree %>%
    group_by(tip.label) %>%
    mutate(n_es = sum(n = n())) %>%
    ungroup()  # Count the number of effect sizes for each species
d.tree <- mutate(d.tree, log_es = log10(n_es))  # Log the number of effect sizes, if needed for visualisations
d.tree_plot <- distinct(d.tree, tip.label, .keep_all = TRUE)  # Only keep one row per species
d.tree_plot <- as.data.frame(d.tree_plot)  # Convert tibble to data frame
d.tree_plot$species_class <- as.character(d.tree_plot$species_class)  # Convert phylum to character
d.tree_plot <- select(d.tree_plot, tip.label, species_class, dispersal_type, n_es, log_es)  # important: select only the columns you need, otherwise this does not run

tree_for_plot <- drop.tip(mytree, setdiff(mytree$tip.label, mytree$tip.label))  # Make sure to keep all matching observations


design_dat <- d.tree %>%
    group_by(tip.label) %>%
    summarise(initial = dispersal_type == "natal")  # Summarise, for each species, whether they tested initial effects, persistent effects, CTmax, or LT50

design_dat <- distinct(design_dat)  # Only keep unique rows
design_dat$initial = as.numeric(design_dat$initial)  # convert TRUE/FALSE to binary values
design_dat$persistent = as.numeric(design_dat$persistent)  # convert TRUE/FALSE to binary values
design_dat$CTmax = as.numeric(design_dat$CTmax)  #convert TRUE/FALSE to binary values
design_dat$LT50 = as.numeric(design_dat$LT50)  #convert TRUE/FALSE to binary values

design_dat <- design_dat %>%
    group_by(tip.label) %>%
    summarise(initial = sum(initial), persistent = sum(persistent), CTmax = sum(CTmax),
        LT50 = sum(LT50))  # calculate the sum for each species (i.e. if 1, the species has the given design or metric)   

# In some cases, values were repeated, so need to replace '2' values by '1'
design_dat$initial[design_dat$initial == "2"] <- "1"
design_dat$persistent[design_dat$persistent == "2"] <- "1"
design_dat$CTmax[design_dat$CTmax == "2"] <- "1"

design_dat$initial = as.factor(design_dat$initial)  # convert back to factor for the plot
design_dat$persistent = as.factor(design_dat$persistent)
design_dat$CTmax = as.factor(design_dat$CTmax)
design_dat$LT50 = as.factor(design_dat$LT50)


design_dat <- mutate(design_dat, persistence = ifelse(initial == "1" & persistent ==
    "1", "both", ifelse(initial == "1" & persistent == "0", "initial", "persistent")))  # If both initial and persistent, indicate 'both', if only initial, indicate 'initial', otherwise indicate 'persistent'

design_dat <- mutate(design_dat, metrics = ifelse(CTmax == "1" & LT50 == "1", "both",
    ifelse(CTmax == "1" & LT50 == "0", "CTmax", "LT50")))  # If both CTmax and LT50, indicate 'both', if only CTmax, indicate 'CTmax', otherwise indicate 'LT50'

d.tree_plot <- left_join(d.tree_plot, design_dat, by = "tip.label")  # Join this information to the rest of the data we want to plot



p <- ggtree(tree_for_plot, layout = "circular", lwd = 0.75)  # Circular tree
p <- p %<+% d.tree_plot  # link plot to data
p2 <- p + geom_fruit(geom = geom_tile, mapping = aes(fill = metrics), width = 0.07,
    offset = 0.085, col = "gray30") + scale_fill_manual(values = c("gray70", "white",
    "black"))  # Create tiles to indicate which metric was used for this species

p3 <- p2 + new_scale_fill() + geom_fruit(geom = geom_tile, mapping = aes(fill = persistence),
    offset = 0.1, width = 0.07, col = "gray30") + scale_fill_manual(values = c("#D95F02",
    "#7570B3", "#1B9E77"))  # Create tiles to indicate whether initial or persistent effects were assessed

d.tree_plot$n_es = as.numeric(d.tree_plot$n_es)  # Convert to numeric 

p4 <- p3 + new_scale_fill() + geom_fruit(geom = geom_bar, mapping = aes(x = n_es,
    fill = class), stat = "identity", col = "gray1", orientation = "y", axis.params = list(axis = "x",
    text.angle = -45, hjust = 0, text.size = 3), border = 1.2, grid.params = list(alpha = 0.35),
    offset = 0.085, pwidth = 0.55, alpha = 0.8) + scale_fill_manual(values = c("#00BBDB",
    "darkslategray4", "darkorange", "chartreuse", "indianred2", "lightsteelblue2",
    "darkorchid", "darkseagreen1", "deeppink", "darkred", "forestgreen", "darkolivegreen2",
    "#39568CFF", "darkgoldenrod2"))  # Display number of effect sizes

p4

 
 
 write.tree(mytree, file = "~/Documents/Files/Post-docs/UNSW 2022-2024/Aim 1/Fitness-and-dispersal-MA/data/species_tree_plotting.tre") #save the tree 
 mytree <- read.tree(file = "~/Documents/Files/Post-docs/UNSW 2022-2024/Aim 1/Fitness-and-dispersal-MA/data/species_tree_plotting.tre") #if you need to read the tree
 
 head(mytree)
 
 
 
 
 
 
 
 
 
 
 
 
 
 ggtree(mytree, layout="circular") +                    
   geom_tiplab( ) 
   

 
  
ggtree(mytree, layout = "circular", branch.length = 'none') %<+% sample_data + # %<+% adds dataframe with sample data to tree 
   new_scale_color() +                             # allows to add an additional color scheme for another variable
     geom_tippoint(
       mapping = aes(color = n),  
       size = 1) +
   scale_colour_gradient(low = "yellow", high = "red") +
   geom_tiplab(                             # adds name of sample to tip of its branch 
     color = 'black',                       
     offset = 1,
     size = 1,
     geom = "text",
     align = TRUE) 
   
 
ggtree(mytree, layout="circular", branch.length = "none") # simple circular tree with all tips aligned
 
ggsave("species_tree_circular_v0.pdf", width = 12, height = 10, scale = 10)
 
ggtree(mytree, layout = "circular", branch.length = 'none') %<+% sample_data + # %<+% adds dataframe with sample data to tree 
   new_scale_color() +                             # allows to add an additional color scheme for another variable
     geom_tippoint(
       mapping = aes(color = n),  
       size = 1) +
   scale_colour_gradient(low = "yellow", high = "red") +
   geom_tiplab(                             # adds name of sample to tip of its branch 
     color = 'black',                       
     offset = 1,
     size = 1,
     geom = "text",
     align = TRUE)  
 
 
 
sample_data <- count(Species_info, common_name)
sample_data <- as.data.frame(sample_data)
sample_data2 <-mutate(sample_data, tip.label = common_name)
mytree$tip.label <- strip_ott_ids(mytree$tip.label, remove_underscores = TRUE)
 summary(sample_data2$tip.label)
 str(sample_data2)
 summary(sample_data2)


#############
#next step is to add the sample size first
#make the text italics
#can add four main silhouttes - mammals, birds, fish, insects (four colours)
#############


# r package: https://cran.r-project.org/web/packages/rphylopic/index.html

img <- pick_phylopic(name = "Canis lupus", n = 4, view = 4)
uuid <- get_uuid(name = "Canis lupus", n = 4)
add_phylopic_base(uuid = uuid, x = 1, y = 1, ysize = 0.25)


p <- p + add_phylopic(uuid = uuid, x = 1, y = 1, ysize = 0.25)
p
